src/witan/models/dem/ccm/core/projection_loop.clj

```diff
@@ -91,6 +91,24 @@
                           (ds/select-columns [:gss-code :sex :age :year :popn]))]
     {:latest-yr-popn survived-popn}))
 
+(defworkflowfn apply-migration
+  "Takes in a dataset of popn estimates for the current year and a
+   dataset with the net migrants for the same year.
+   Returns a dataset where the migrants are added to the popn dataset"
+  {:witan/name :ccm-core/apply-migration
+   :witan/version "1.0"
+   :witan/input-schema {:latest-yr-popn PopulationSchema :net-migration NetMigrationSchema}
+   :witan/param-schema {:_ nil}
+   :witan/output-schema {:latest-yr-popn PopulationSchema}}
+  [{:keys [latest-yr-popn net-migration]} _]
+  (let [migration-ds (ds/select-columns net-migration [:gss-code :sex :age :net-mig])
+        popn-mig (i/$join [[:gss-code :sex :age] [:gss-code :sex :age]]
+                          latest-yr-popn migration-ds)
+        popn-w-migrants (-> (i/replace-column :popn (i/$map (fn [popn mig] (+ popn mig))
+                                                            [:popn :net-mig] popn-mig) popn-mig)
+                            (ds/select-columns [:gss-code :sex :age :year :popn]))]
```

Comment:
105 -> 109 is repeated in apply-deaths. As in the code joins a dataset, then combines it. You could instead join both sets and then in one pass add migrations and subtract deaths. Saving another scan per loop. There might be a way of reducing the join costs if we know that the data is ordered, depends on the production of those values a bit, but it might be worth the ordering cost up front, if then for each loop we can just zip the data together.